vyper/ast.py

```diff
@@ -0,0 +1,278 @@
+from itertools import (
+    chain,
+)
+from typing import (
+    Any,
+    List as ListTyping,
+)
+
+from vyper.exceptions import (
+    CompilerPanic,
+)
+
+
+class VyperNode:
+    __slots__ = ('node_id', 'source_code', 'col_offset', 'lineno')
+    ignored_fields = ['ctx', ]
+    only_empty_fields: ListTyping[Any] = []
+
+    @classmethod
+    def get_slots(cls):
+        return chain.from_iterable(
+            getattr(klass, '__slots__', [])
+            for klass in cls.__class__.mro(cls)
+        )
+
+    def __init__(self, **kwargs):
+
+        for field_name, value in kwargs.items():
+            if field_name in self.get_slots():
+                setattr(self, field_name, value)
+            elif value:
+                raise CompilerPanic(
+                    f'Unsupported non-empty value field_name: {field_name}, '
+                    f' class: {type(self)} value: {value}'
+                )
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            for field_name in self.get_slots():
+                if field_name not in ('node_id', 'source_code', 'col_offset', 'lineno'):
```

Comment:
This list of field names is probably best kept in a module level constant (sighly more performant and easier to read)

Reply:
@jacqueswww is this resolved?

Reply:
Yip it is, created BASE_NODE_ATTRIBUTES. Just missed to resolve this ;)